iocore/net/NetTimeout.h

```diff
@@ -0,0 +1,253 @@
+  void cancel_inactive_timeout();
+  bool is_active_timeout_expired(ink_hrtime now);
+  bool is_inactive_timeout_expired(ink_hrtime now);
+  void update_inactivity();
+  ink_hrtime active_timeout_in() const;
+  ink_hrtime inactive_timeout_in() const;
+  ink_hrtime next_active_timeout_at() const;
+  ink_hrtime next_inactive_timeout_at() const;
+
+private:
+  ink_hrtime _active_timeout_in        = 0;
+  ink_hrtime _inactive_timeout_in      = 0;
+  ink_hrtime _next_active_timeout_at   = 0;
+  ink_hrtime _next_inactive_timeout_at = 0;
+};
+
+/**
+  ActivityCop - Check activity of T in the List in every @f seconds
+
+  T have to handle VC_EVENT_ACTIVE_TIMEOUT and VC_EVENT_INACTIVITY_TIMEOUT events.
+
+  TODO: add concepts like below with C++20
+  ```
+  template <class T, class List = DLL<T>>
+  concept Timeoutable = requires(T *t, List *list, ink_hrtime time) {
+    t->handleEvent();
+    t->head;
+    {list->next(t)} -> std::convertible_to<T *>;
+    {t->is_active_timeout_expired(time)} -> std::same_as<bool>;
+    {t->is_inactive_timeout_expired(time)} -> std::same_as<bool>;
+  };
+  ```
+ */
+template <class T, class List = DLL<T>> class ActivityCop : Continuation
+{
+public:
+  ActivityCop(){};
+  ActivityCop(Ptr<ProxyMutex> &m, List *l, int f);
+
+  void start();
+  void stop();
+  int check_activity(int event, Event *e);
+
+private:
+  Event *_event = nullptr;
+  List *_list   = nullptr;
+  int _freq     = 1;
+};
+
+////
+// Inline functions
+
+//
+// NetTimeout
+//
+inline void
+NetTimeout::set_active_timeout(ink_hrtime timeout_in)
+{
+  if (timeout_in == 0) {
+    return;
+  }
+
+  _active_timeout_in      = timeout_in;
+  _next_active_timeout_at = Thread::get_hrtime() + timeout_in;
+}
+
+inline void
+NetTimeout::set_inactive_timeout(ink_hrtime timeout_in)
+{
+  if (timeout_in == 0) {
+    return;
+  }
+
+  _inactive_timeout_in      = timeout_in;
+  _next_inactive_timeout_at = Thread::get_hrtime() + timeout_in;
+}
+
+inline void
+NetTimeout::cancel_active_timeout()
+{
+  _active_timeout_in      = 0;
+  _next_active_timeout_at = 0;
+}
+
+inline void
+NetTimeout::cancel_inactive_timeout()
+{
+  _inactive_timeout_in      = 0;
+  _next_inactive_timeout_at = 0;
+}
+
+inline void
+NetTimeout::update_inactivity()
+{
+  if (_inactive_timeout_in == 0) {
+    return;
+  }
+
+  _next_inactive_timeout_at = Thread::get_hrtime() + _inactive_timeout_in;
+}
+
+inline bool
+NetTimeout::is_active_timeout_expired(ink_hrtime now)
+{
+  ink_assert(now > 0);
+
+  if (_active_timeout_in == 0) {
+    return false;
+  }
+
+  if (0 < _next_active_timeout_at && _next_active_timeout_at < now) {
+    Debug("activity_cop", "active timeout cont=%p now=%" PRId64 " timeout_at=%" PRId64 " timeout_in=%" PRId64, this,
+          ink_hrtime_to_sec(now), ink_hrtime_to_sec(_next_active_timeout_at), ink_hrtime_to_sec(_active_timeout_in));
+    return true;
+  }
+
+  return false;
+}
+
+inline bool
+NetTimeout::is_inactive_timeout_expired(ink_hrtime now)
```

Comment:
Why is this check done by NetTimeout itself?

The interaction between ActivityCop and NetTimeout looks weird.
AC: Hey, it's March 1st. Is your visa expired?
NT: Yes, it is.
AC: Your visa is expired.
NT: (I know, I told you.)

I'd do
AC: Hey, when is your visa expire date?
NT: It's March 1st.
AC: Your visa is expired.

Reply:
The main motivation is hiding private members as much as possible. Maybe I did too much.

Reply:
I revisit this but it looks like what AC needs to know is timeout is expired or not.
We see a similar code & relations for checking a connection is closed or not.

Reply:
Just asking expiration time is enough to know whether a stream is timed out. I don't see any needs to make two function calls. I'd move the check logic to AC and make NetTimeout to a simple data structure (with accessors).